server/datastore/mysql/mysql.go

```diff
@@ -50,6 +51,40 @@ func (d *Datastore) getTransaction(opts []kolide.OptionalArg) dbfunctions {
 	return result
 }
 
+type txFn func(*sqlx.Tx) error
+
+// withRetryTxx provides a common way to commit/rollback a txFn wrapped in a retry with exponential backoff
+func (d *Datastore) withRetryTxx(fn txFn) (err error) {
+	operation := func() error {
+		tx, err := d.db.Beginx()
+		if err != nil {
+			return errors.Wrap(err, "creating transaction")
+		}
+
+		err = fn(tx)
+
+		defer func() {
+			if p := recover(); p != nil {
+				tx.Rollback()
+				panic(p)
+			} else if err != nil {
+				rbErr := tx.Rollback()
+				if rbErr != nil && rbErr != sql.ErrTxDone {
+					panic(fmt.Sprintf("got err '%s' rolling back after err '%s'", rbErr, err))
+				}
+			} else {
+				err = tx.Commit()
+			}
+		}()
+
+		return err
+	}
```

Comment:
I think this `defer` needs to come before calling the `fn`. Otherwise we miss catching any panics that happen in the function.

```suggestion
		defer func() {
			if p := recover(); p != nil {
				tx.Rollback()
				panic(p)
			} else if err != nil {
				rbErr := tx.Rollback()
				if rbErr != nil && rbErr != sql.ErrTxDone {
					panic(fmt.Sprintf("got err '%s' rolling back after err '%s'", rbErr, err))
				}
			} else {
				err = tx.Commit()
			}
		}()

		return err
	}
	
	err = fn(tx)
```

Reply:
Ah, good call. I think we'll need to split it into two deferred calls then, otherwise `err` will only contain results from the d.db.Beginx() call.

Reply:
Pushed up what I think it should look like by only handling the panic in the defer.